Azure/sonic-buildimage
# libyang python APIs

**- What I did**
        libyang Python APIs:
          - python APIs based on libyang
          - functions to load/merge yang models and Yang data files
          - function to validate data trees based on Yang models
          - functions to merge yang data files/trees
          - add/set/delete node in schema and data trees
          - find data/schema nodes from xpath from the Yang data/schema tree in memory 
          - find dependencies
          - dump the data tree in json/xml 

**- How I did it**
     -  based on libyang python2 modules which is built by libyang
     -  add functions to load/validate Yang models and Yang data trees
     -  Add functions to 
    
**- How to verify it**
    -  pytests while build sonic-yang-mgmt python wheel package
    -  create sample Yang models for sonic ACL/interface/port/clan
    -  sample Yang data files
    -  test file with test cases,  it includes categories to test, xpath and expected results for testing Yang data/schema tree wrapper functions

pytests while build target/python-wheels/sonic_yang_mgmt-1.0-py2-none-any.whl:

============================= test session starts ==============================
platform linux2 -- Python 2.7.13, pytest-3.0.6, py-1.4.32, pluggy-0.4.0
rootdir: /sonic/src/sonic-yang-mgmt, inifile:
plugins: cov-2.4.0
collected 18 items

tests/test_sonic_yang_mgmt.py .
tests/libyang-python-tests/test_sonic_yang.py .................

========================== 18 passed in 0.21 seconds ===========================

**- Description for the changelog**


Note:    Due to dependency on PR#3861 (https://github.com/Azure/sonic-buildimage/pull/3861), this PR has 2 commits, please review the 2nd commit only(the first commit in this PR is a cherry-pick of PR#3861).
      




**- A picture of a cute animal (not mandatory but encouraged)**
